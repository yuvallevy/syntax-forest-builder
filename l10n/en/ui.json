{
  "toolbox": {
    "title": "Tools",
    "items": {
      "undo": {
        "title": "Undo",
        "description": "Undo the last action."
      },
      "redo": {
        "title": "Redo",
        "description": "Redo the last undone action."
      },
      "add": {
        "title": "Add",
        "description": "Add a new parent node for the selected text or nodes."
      },
      "delete": {
        "title": "Delete",
        "descriptions": {
          "deleteNodes": "Delete the selected nodes.",
          "deleteTrees": "Delete the selected trees."
        }
      },
      "edit": {
        "title": "Edit",
        "description": "Edit the selected node."
      },
      "triangle": {
        "title": "Triangle",
        "description": "Toggle triangle connectors for the selected terminal nodes."
      },
      "adopt": {
        "title": "Adopt",
        "description": "Adopt one or more nodes as children of the selected node."
      },
      "disown": {
        "title": "Disown",
        "description": "Disown one or more children of the selected node."
      },
      "resetPosition": {
        "title": "Reset position",
        "description": "Relocate the selected nodes to their original positions."
      },
      "selectTrees": {
        "title": "Select trees",
        "description": "Select entire trees instead of individual nodes."
      }
    },
    "tooltips": {
      "hotkeyHold": "hold {{hotkey}}"
    }
  },
  "mainMenu": {
    "file": {
      "title": "File",
      "items": {
        "open": "Open",
        "save": "Save",
        "saveAs": "Save As",
        "currentlyOpenFile": "Currently open file:"
      }
    },
    "settings": {
      "title": "Settings"
    },
    "about": {
      "title": "About"
    }
  },
  "fileIo": {
    "title": {
      "save": "Save",
      "open": "Open"
    },
    "listHeaders": {
      "name": "Name",
      "size": "Size",
      "modified": "Modified"
    },
    "labels": {
      "fileName": "Save as:"
    },
    "buttons": {
      "save": "Save"
    }
  },
  "settings": {
    "items": {
      "autoFormatBracketsAsSubscript": {
        "title": "Format brackets/parentheses as subscript",
        "description": "When this is enabled, letters and numbers in parentheses or brackets such as (i) and [i] will automatically be rendered as subscript. (Does not work on all letters)"
      },
      "experimentalTextWidth": {
        "title": "Use experimental node positioning method (slightly slower, more accurate)"
      }
    },
    "buttons": {
      "done": "Done"
    }
  }
}
