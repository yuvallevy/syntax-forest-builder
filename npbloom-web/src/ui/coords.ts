import { CoordsInPlot, PositionedNode, PositionedTree } from 'npbloom-core';

type ClientX = number;
type ClientY = number;
type DClientX = number;
type DClientY = number;

export type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientCoordsOffset = {
  dClientX: DClientX;
  dClientY: DClientY;
}

export type ClientRect = {
  topLeft: ClientCoords;
  bottomRight: ClientCoords;
};

export type PlotRect = {
  topLeft: CoordsInPlot;
  bottomRight: CoordsInPlot;
};

// Calculate plot coords from client coords & vice versa, when they're different (when zoom and/or pan are implemented)

export const plotCoordsToClientCoords = (plotCoords: CoordsInPlot): ClientCoords => ({
  clientX: plotCoords.plotX,
  clientY: plotCoords.plotY,
});

export const clientCoordsToPlotCoords = (clientCoords: ClientCoords): CoordsInPlot => new CoordsInPlot(
  clientCoords.clientX,
  clientCoords.clientY,
);

export const clientRectToPlotRect = (clientRect: ClientRect): PlotRect => ({
  topLeft: clientCoordsToPlotCoords(clientRect.topLeft),
  bottomRight: clientCoordsToPlotCoords(clientRect.bottomRight),
});

/**
 * Returns the center coordinate of the given node in the given tree with respect to its containing plot.
 */
export const calculateNodeCenterOnPlot = (tree: PositionedTree) => (node: PositionedNode): CoordsInPlot => new CoordsInPlot(
  tree.position.plotX + node.position.treeX,
  tree.position.plotY + node.position.treeY - 9,
);
