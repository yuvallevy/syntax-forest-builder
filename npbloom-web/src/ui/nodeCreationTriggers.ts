import { Id, Sentence, StringSlice } from '../content/types';
import {
  getTopLevelPositionedNodes, isSliceUnassigned, sortPositionedNodesByXCoord
} from '../content/positioned/positionedEntityHelpers';
import { windowed } from '../util/objTransforms';
import {
  determineNaturalParentNodePosition, sliceOffsetAndWidth, StrWidthFunc
} from '../content/positioned/positioning';
import { PositionedTree, PositionInTree } from '../content/positioned/types';

const MAX_TRIGGER_WIDTH = 32;
const MAX_TRIGGER_PADDING_TOP = 28;
const MAX_TRIGGER_PADDING_BOTTOM = 20;

type ChildIdsOrSlice = { childIds: Id[], childPositions: PositionInTree[] } | { slice: StringSlice };

type NodeCreationTarget = ChildIdsOrSlice & {
  position: PositionInTree;
};

export type NodeCreationTrigger = ChildIdsOrSlice & {
  origin: PositionInTree;
  topLeft: PositionInTree;
  bottomRight: PositionInTree;
};

const getWordSlices = (sentence: Sentence, wordRegex?: RegExp): StringSlice[] => {
  const _wordRegex = wordRegex || /['A-Za-z\u00c0-\u1fff]+/g;
  const match = _wordRegex.exec(sentence);
  if (!match) return [];
  const sliceStart = match.index;
  const sliceEnd = sliceStart + match[0].length;
  const nextSlices = getWordSlices(sentence, _wordRegex);  // The next time exec is called it will return the next word
  return [[sliceStart, sliceEnd], ...nextSlices];
};

const getNodeCreationTargetsForTree = (strWidthFunc: StrWidthFunc) => (positionedTree: PositionedTree): NodeCreationTarget[] => {
  // We only need node creation triggers to add parent nodes for top-level nodes, so discard the rest
  const topLevelNodes = getTopLevelPositionedNodes(positionedTree);
  const topLevelNodeIds = sortPositionedNodesByXCoord(positionedTree)(Object.keys(topLevelNodes));

  // We also need one trigger above each space between two horizontally adjacent nodes
  const topLevelNodeIdPairs = windowed(topLevelNodeIds, 2);

  // Finally, we need one trigger for each word that isn't already assigned to a terminal node
  const unassignedSlices = getWordSlices(positionedTree.sentence).filter(isSliceUnassigned(positionedTree));

  // Find the targets for all of these triggers, i.e. where nodes can be added
  const parentNodeCreationTargets: NodeCreationTarget[] = topLevelNodeIds.map(id => [id]).concat(topLevelNodeIdPairs)
    .map(nodeIds => ({
      position: determineNaturalParentNodePosition(nodeIds.map(nodeId => positionedTree.nodes[nodeId].position)),
      childIds: nodeIds,
      childPositions: nodeIds.map(nodeId => positionedTree.nodes[nodeId].position),
    }));
  const terminalNodeCreationTargets: NodeCreationTarget[] = unassignedSlices.map(slice => {
    const [widthBeforeSlice, sliceWidth] = sliceOffsetAndWidth(strWidthFunc)(positionedTree.sentence)(slice);
    return {
      position: { treeX: widthBeforeSlice + (sliceWidth / 2), treeY: -MAX_TRIGGER_PADDING_BOTTOM },
      slice,
    };
  });

  return parentNodeCreationTargets.concat(terminalNodeCreationTargets);
};

export const getNodeCreationTriggersForTree =
  (strWidthFunc: StrWidthFunc) =>
  (positionedTree: PositionedTree): NodeCreationTrigger[] =>
    getNodeCreationTargetsForTree(strWidthFunc)(positionedTree).map(target => ({
      origin: target.position,
      topLeft: {
        treeX: target.position.treeX - MAX_TRIGGER_WIDTH / 2,
        treeY: target.position.treeY - MAX_TRIGGER_PADDING_TOP,
      },
      bottomRight: {
        treeX: target.position.treeX + MAX_TRIGGER_WIDTH / 2,
        treeY: target.position.treeY + MAX_TRIGGER_PADDING_BOTTOM,
      },
      ...(
        'childIds' in target
          ? { childIds: target.childIds, childPositions: target.childPositions }
          : { slice: target.slice }
      ),
    }));
