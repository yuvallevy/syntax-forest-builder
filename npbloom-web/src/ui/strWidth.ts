import { mockStrWidth } from 'npbloom-core';

// Overwrite this with an actual string width measuring function when the time comes
const strWidth = mockStrWidth;

export default strWidth;
