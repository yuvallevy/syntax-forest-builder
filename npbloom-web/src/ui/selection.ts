import { union } from '../util/objTransforms';
import { Id } from '../types';
import {
  CoordsInPlot, NodeIndicatorInPlot, PositionedNode, PositionedTree, StringSlice, UnpositionedPlot
} from 'npbloom-core';
import { calculateNodeCenterOnPlot, PlotRect } from './coords';

export type NodeSelectionInPlot = { nodeIndicators: NodeIndicatorInPlot[] };
export type SliceSelectionInPlot = { treeId: Id, slice: StringSlice };
export type SelectionInPlot = NodeSelectionInPlot | SliceSelectionInPlot;

export const isNodeSelection = (selection: SelectionInPlot): selection is NodeSelectionInPlot => 'nodeIndicators' in selection;
export const isSliceSelection = (selection: SelectionInPlot): selection is SliceSelectionInPlot => 'slice' in selection;

export type NodeSelectionAction = 'select' | 'adopt' | 'disown';
export type NodeSelectionMode = 'set' | 'add';

export const applySelection = (
  mode: NodeSelectionMode, newNodeIndicators: NodeIndicatorInPlot[], existingNodeIndicators?: NodeIndicatorInPlot[]) =>
  mode === 'add' ? union(existingNodeIndicators || [], newNodeIndicators)
    : newNodeIndicators;

/**
 * Returns whether the node indicated by the given indicator exists in the given plot.
 */
const indicatorTargetExistsInPlot = (plot: UnpositionedPlot) => (nodeIndicator: NodeIndicatorInPlot) =>
  plot.contains(nodeIndicator.treeId) && plot.tree(nodeIndicator.treeId).contains(nodeIndicator.nodeId);

/**
 * Returns a copy of the given node selection including only nodes matching the given predicate.
 */
const filterNodesInSelection = (
  selection: NodeSelectionInPlot,
  predicate: (nodeIndicator: NodeIndicatorInPlot) => boolean,
): SelectionInPlot =>
  ({ nodeIndicators: selection.nodeIndicators.filter(predicate) });

/**
 * Removes nonexistent nodes from the given selection, based on the given plot.
 */
export const pruneSelection = (selection: SelectionInPlot, plot: UnpositionedPlot): SelectionInPlot =>
  isSliceSelection(selection)
    ? (plot.contains(selection.treeId) ? selection : { nodeIndicators: [] })
    : filterNodesInSelection(selection, indicatorTargetExistsInPlot(plot));

/**
 * Returns whether the given point is inside the given rectangle.
 */
const isPointInPlotRect = (rect: PlotRect) => (coords: CoordsInPlot) =>
  coords.plotX >= rect.topLeft.plotX && coords.plotY >= rect.topLeft.plotY &&
  coords.plotX <= rect.bottomRight.plotX && coords.plotY <= rect.bottomRight.plotY;

export const isNodeInRect = (rect: PlotRect) => (tree: PositionedTree) => (node: PositionedNode) =>
  isPointInPlotRect(rect)(calculateNodeCenterOnPlot(tree)(node));
