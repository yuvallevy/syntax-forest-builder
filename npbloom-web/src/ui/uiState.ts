import {
  AddPlot, AddTree, AdoptNodes, arrayFromSet, contentReducer, DeleteNodes, DeletePlot, DisownNodes, DeleteTree,
  initialContentState, InsertedBranchingNode, InsertedTerminalNode, InsertNode, newNodeFromSelection,
  NodeIndicatorInPlot, NodeSelectionAction, NodeSelectionInPlot, MoveNodes, PlotCoordsOffset, pruneSelection, Redo,
  ResetNodePositions, ResetPlot, SelectionInPlot, set, SetNodeLabel, SetTriangle, SetSentence, SliceSelectionInPlot,
  sortNodesByXCoord, StringSlice, TreeCoordsOffset, Undo, UnpositionedBranchingNode, UnpositionedPlot,
  UnpositionedTerminalNode
} from 'npbloom-core';
import { Id, NodeLabel, PlotIndex, Sentence, UndoableContentState } from '../types';
import strWidth from './strWidth';
import { without } from '../util/objTransforms';

export type UiAction =
  | { type: 'setActivePlotIndex', newPlotIndex: PlotIndex }
  | { type: 'addPlot' }
  | { type: 'deletePlot', plotIndex: PlotIndex }
  | { type: 'setSelection', newSelection: SelectionInPlot }
  | { type: 'selectParentNodes' }
  | { type: 'selectChildNode', side: 'left' | 'right' | 'center' }
  | { type: 'startEditing' }
  | { type: 'stopEditing' }
  | { type: 'setEditedNodeLabel', newLabel: NodeLabel }
  | { type: 'setSelectionAction', selectionAction: NodeSelectionAction }
  | { type: 'addNodeBySelection', newNodeId: Id }
  | { type: 'addNodeByTarget', treeId: Id, newNodeId: Id, targetChildIds: Id[] }
  | { type: 'addNodeByTarget', treeId: Id, newNodeId: Id, targetSlice: StringSlice, triangle: boolean }
  | { type: 'deleteSelectedNodes' }
  | { type: 'adoptNodesBySelection', adoptedNodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'disownNodesBySelection', disownedNodeIndicators: NodeIndicatorInPlot[] }
  | { type: 'moveSelectedNodes', dx: number, dy: number }
  | { type: 'resetSelectedNodePositions' }
  | { type: 'toggleTriangle' }
  | { type: 'setSentence', newSentence: Sentence, oldSelectedSlice: StringSlice, treeId?: Id }
  | { type: 'addTree', newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', treeId: Id }
  | { type: 'undo' }
  | { type: 'redo' }
  ;

export type UiState = {
  contentState: UndoableContentState;
  activePlotIndex: PlotIndex;
  selection: SelectionInPlot;
  selectionAction: NodeSelectionAction;
  editedNodeIndicator?: NodeIndicatorInPlot;
};

export const initialUiState: UiState = {
  activePlotIndex: 0,
  contentState: initialContentState.get(),
  selection: new NodeSelectionInPlot(),
  selectionAction: NodeSelectionAction.Select,
};

const selectParentNodes = (activePlot: UnpositionedPlot, selection: SelectionInPlot): NodeSelectionInPlot => {
  if (selection instanceof SliceSelectionInPlot) {
    return new NodeSelectionInPlot(
      set(arrayFromSet<Id>(activePlot.tree(selection.treeId).getNodeIdsAssignedToSlice(selection.slice))
        .map(nodeId => new NodeIndicatorInPlot(selection.treeId, nodeId)))
    );
  } else if (selection instanceof NodeSelectionInPlot) {
    if (selection.nodeIndicators.length === 0) return selection;
    return new NodeSelectionInPlot(activePlot.getParentNodeIds(selection.nodeIndicators));
  } else {
    return new NodeSelectionInPlot();
  }
};

export const uiReducer = (state: UiState, action: UiAction): UiState => {
  const activePlot = state.contentState.current.plots[state.activePlotIndex];
  const selectedTreeId =
    state.selection instanceof SliceSelectionInPlot ? state.selection.treeId
      : state.selection instanceof NodeSelectionInPlot && arrayFromSet(state.selection.nodeIndicators).length > 0
      ? arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators)[0].treeId
      : undefined;
  switch (action.type) {
    case 'setActivePlotIndex':
      return {
        ...state,
        activePlotIndex: action.newPlotIndex,
        selection: new NodeSelectionInPlot(),
        selectionAction: NodeSelectionAction.Select,
        editedNodeIndicator: undefined,
      };
    case 'addPlot':
      return {
        ...state,
        contentState: contentReducer(state.contentState, AddPlot.getInstance()),
        activePlotIndex: state.contentState.current.plots.length,
        selection: new NodeSelectionInPlot(),
        selectionAction: NodeSelectionAction.Select,
        editedNodeIndicator: undefined,
      };
    case 'deletePlot': {
      const isLastRemainingPlot = state.contentState.current.plots.length === 1;
      const newContentState = contentReducer(state.contentState,
        isLastRemainingPlot ? new ResetPlot(action.plotIndex) : new DeletePlot(action.plotIndex));
      const newActivePlotIndex = state.activePlotIndex < newContentState.current.plots.length ? state.activePlotIndex
        : newContentState.current.plots.length - 1;
      return {
        ...state,
        contentState: newContentState,
        activePlotIndex: newActivePlotIndex,
        selection: new NodeSelectionInPlot(),
        selectionAction: NodeSelectionAction.Select,
        editedNodeIndicator: undefined,
      };
    }
    case 'setSelection': {
      return {
        ...state,
        selection: action.newSelection,
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'selectParentNodes': {
      const parentSelection = selectParentNodes(activePlot, state.selection);
      return {
        ...state,
        selection: parentSelection,
        editedNodeIndicator: state.editedNodeIndicator
          ? arrayFromSet<NodeIndicatorInPlot>(parentSelection.nodeIndicators)[0] : undefined,
      };
    }
    case 'selectChildNode': {
      if (
        !(state.selection instanceof NodeSelectionInPlot) ||  // no nodes selected
        (state.selection.nodeIndicators.length > 1) ||  // multiple nodes selected
        !selectedTreeId  // could not figure out tree ID for some other reason
      ) return state;
      const selectedNodeObject = activePlot.tree(selectedTreeId)
        .node(arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators)[0].nodeId);
      if (!(selectedNodeObject instanceof UnpositionedBranchingNode)) return state;

      const selectedNodeChildren = arrayFromSet<Id>(selectedNodeObject.children);
      const childNodesSortedByX =
        sortNodesByXCoord(strWidth, activePlot.tree(selectedTreeId), selectedNodeObject.children);

      const childSelection: NodeSelectionInPlot | undefined =
        action.side === 'center' && selectedNodeChildren.length === 1 ?
          new NodeSelectionInPlot(set([new NodeIndicatorInPlot(selectedTreeId, selectedNodeChildren[0])])) :
        action.side === 'center' && selectedNodeChildren.length >= 3 ?
          new NodeSelectionInPlot(set([new NodeIndicatorInPlot(selectedTreeId, childNodesSortedByX[1])])) :
        action.side !== 'center' && selectedNodeChildren.length >= 2 ?
          new NodeSelectionInPlot(
            set([
              new NodeIndicatorInPlot(
                selectedTreeId,
                childNodesSortedByX[action.side === 'left' ? 0 : (selectedNodeChildren.length - 1)]
              )
            ])
          ) : undefined;

      if (!childSelection) return state;
      return {
        ...state,
        selection: childSelection,
        editedNodeIndicator: state.editedNodeIndicator
          ? arrayFromSet<NodeIndicatorInPlot>(childSelection.nodeIndicators)[0] : undefined,
      };
    }
    case 'startEditing': {
      if (!(state.selection instanceof NodeSelectionInPlot) ||
        arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators).length !== 1) return state;
      return {
        ...state,
        editedNodeIndicator: arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators)[0],
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'stopEditing': {
      return {
        ...state,
        editedNodeIndicator: undefined,
      };
    }
    case 'setEditedNodeLabel': {
      if (!state.editedNodeIndicator) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new SetNodeLabel(
          state.activePlotIndex,
          state.editedNodeIndicator,
          action.newLabel,
        )),
      };
    }
    case 'setSelectionAction': {
      return { ...state, selectionAction: action.selectionAction };
    }
    case 'addNodeBySelection': {
      if (!selectedTreeId) return state;
      const newNodeIndicator = new NodeIndicatorInPlot(selectedTreeId, action.newNodeId);
      return {
        ...state,
        contentState: contentReducer(state.contentState, new InsertNode(
          state.activePlotIndex,
          selectedTreeId,
          action.newNodeId,
          newNodeFromSelection(state.selection, activePlot.tree(selectedTreeId).sentence),
        )),
        selection: new NodeSelectionInPlot(set([newNodeIndicator])),
        selectionAction: NodeSelectionAction.Select,
        editedNodeIndicator: newNodeIndicator,
      };
    }
    case 'addNodeByTarget': {
      const newNodeIndicator = new NodeIndicatorInPlot(action.treeId, action.newNodeId);
      return {
        ...state,
        contentState: contentReducer(state.contentState, new InsertNode(
          state.activePlotIndex,
          action.treeId,
          action.newNodeId,
          'targetChildIds' in action
            ? new InsertedBranchingNode('', null, action.targetChildIds)
            : new InsertedTerminalNode('', null, action.targetSlice, action.triangle),
        )),
        selection: new NodeSelectionInPlot(set([newNodeIndicator])),
        selectionAction: NodeSelectionAction.Select,
        editedNodeIndicator: newNodeIndicator,
      };
    }
    case 'deleteSelectedNodes': {
      if (!(state.selection instanceof NodeSelectionInPlot)) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new DeleteNodes(
          state.activePlotIndex,
          state.selection.nodeIndicators,
        )),
        selection: new NodeSelectionInPlot(set(
          without(
            arrayFromSet(activePlot.getChildNodeIds(state.selection.nodeIndicators)),
            // Currently selected nodes are about to be deleted, so they should not be selected after deletion
            // (this can happen when two deleted nodes are parent and child)
            arrayFromSet(state.selection.nodeIndicators),
          ),
        )),
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'adoptNodesBySelection': {
      if (
        !(state.selection instanceof NodeSelectionInPlot) ||  // no nodes selected
        arrayFromSet(state.selection.nodeIndicators).length > 1 ||  // multiple nodes selected
        !selectedTreeId  // could not figure out tree ID for some other reason
      ) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new AdoptNodes(
          state.activePlotIndex,
          selectedTreeId,
          arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators)[0].nodeId,
          set(action.adoptedNodeIndicators
            .filter(({ treeId }) => treeId === selectedTreeId).map(({ nodeId }) => nodeId)),
        )),
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'disownNodesBySelection': {
      if (
        !(state.selection instanceof NodeSelectionInPlot) ||  // no nodes selected
        arrayFromSet(state.selection.nodeIndicators).length > 1 ||  // multiple nodes selected
        !selectedTreeId  // could not figure out tree ID for some other reason
      ) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new DisownNodes(
          state.activePlotIndex,
          selectedTreeId,
          arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators)[0].nodeId,
          set(action.disownedNodeIndicators
            .filter(({ treeId }) => treeId === selectedTreeId).map(({ nodeId }) => nodeId)),
        )),
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'moveSelectedNodes': {
      if (!(state.selection instanceof NodeSelectionInPlot)) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new MoveNodes(
          state.activePlotIndex,
          state.selection.nodeIndicators,
          new TreeCoordsOffset(action.dx, action.dy),
        )),
      }
    }
    case 'resetSelectedNodePositions': {
      if (!(state.selection instanceof NodeSelectionInPlot)) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new ResetNodePositions(
          state.activePlotIndex,
          state.selection.nodeIndicators,
        )),
      };
    }
    case 'toggleTriangle': {
      if (!(state.selection instanceof NodeSelectionInPlot)) return state;
      const currentlyTriangle = arrayFromSet<NodeIndicatorInPlot>(state.selection.nodeIndicators).every(({ treeId, nodeId }) => {
        const node = activePlot.tree(treeId).node(nodeId);
        return node instanceof UnpositionedTerminalNode ? node.triangle : false;
      });
      return {
        ...state,
        contentState: contentReducer(state.contentState, new SetTriangle(
          state.activePlotIndex,
          state.selection.nodeIndicators,
          !currentlyTriangle,
        )),
      }
    }
    case 'setSentence': {
      if (!selectedTreeId) return state;
      return {
        ...state,
        contentState: contentReducer(state.contentState, new SetSentence(
          state.activePlotIndex,
          action.treeId || selectedTreeId,
          action.newSentence,
          action.oldSelectedSlice,
        )),
      };
    }
    case 'addTree': {
      return {
        ...state,
        contentState: contentReducer(state.contentState, new AddTree(
          state.activePlotIndex,
          action.newTreeId,
          action.offset,
        )),
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'removeTree': {
      return {
        ...state,
        contentState: contentReducer(state.contentState, new DeleteTree(
          state.activePlotIndex,
          action.treeId,
        )),
        selectionAction: NodeSelectionAction.Select,
      };
    }
    case 'undo':
    case 'redo': {
      const newContentState = contentReducer(state.contentState,
        action.type === 'undo' ? Undo.getInstance() : Redo.getInstance());
      const newActivePlotIndex = state.activePlotIndex < newContentState.current.plots.length ? state.activePlotIndex
        : newContentState.current.plots.length - 1;
      return {
        ...state,
        contentState: newContentState,
        activePlotIndex: newActivePlotIndex,
        selection: pruneSelection(state.selection, newContentState.current.plots[newActivePlotIndex]),
      };
    }
  }
};
