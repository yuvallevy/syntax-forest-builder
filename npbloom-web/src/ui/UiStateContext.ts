import { createContext } from 'react';
import { initialUiState, UiAction, UiState } from './uiState';

const UiStateContext = createContext<{
  state: UiState;
  dispatch: React.Dispatch<UiAction>;
}>({
  state: initialUiState,
  dispatch: () => undefined,
});

export default UiStateContext;
