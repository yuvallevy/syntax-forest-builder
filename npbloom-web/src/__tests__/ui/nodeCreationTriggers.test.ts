import { describe, expect, it } from 'vitest';
import {
  CoordsInPlot, CoordsInTree, idMap, PositionedBranchingNode, PositionedTerminalNode, PositionedTree, set, StringSlice,
  TreeXRange
} from 'npbloom-core';
import { getNodeCreationTriggersForTree } from '../../ui/nodeCreationTriggers';
import mockStrWidth from '../__mocks__/mockStrWidth';

describe('node creation triggers', () => {
  const treeWithNoUnassignedSlices = new PositionedTree(
    'Noun verbs very adverbly.',
    idMap({
      'b': new PositionedBranchingNode('NP', new CoordsInTree(18, -60), set(['c'])),
      'c': new PositionedTerminalNode('N', new CoordsInTree(18, -2), new StringSlice(0, 4)),
      'd': new PositionedBranchingNode('VP', new CoordsInTree(89.25, -60), set(['e', 'f'])),
      'e': new PositionedTerminalNode('V', new CoordsInTree(57, -2), new StringSlice(5, 10)),
      'f': new PositionedTerminalNode('AdvP', new CoordsInTree(121.5, -30), new StringSlice(11, 24), new TreeXRange(72, 104)),
    }),
    new CoordsInPlot(50, -32),
    104,
  );
  const treeWithUnassignedSlices = new PositionedTree(
    'Noun verbs very adverbly.',
    idMap({
      'b': new PositionedBranchingNode('NP', new CoordsInTree(18, -60), set(['c'])),
      'c': new PositionedTerminalNode('N', new CoordsInTree(18, -2), new StringSlice(0, 4)),
      'd': new PositionedBranchingNode('VP', new CoordsInTree(57, -60), set(['e'])),
      'e': new PositionedTerminalNode('V', new CoordsInTree(57, -2), new StringSlice(5, 10)),
    }),
    new CoordsInPlot(50, -32),
    104,
  );

  it('generates node creation triggers for a tree with all slices assigned', () => {
    expect(getNodeCreationTriggersForTree(mockStrWidth)(treeWithNoUnassignedSlices)).toMatchSnapshot();
  });

  it('generates node creation triggers for a tree with unassigned slices', () => {
    expect(getNodeCreationTriggersForTree(mockStrWidth)(treeWithUnassignedSlices)).toMatchSnapshot();
  });
});
