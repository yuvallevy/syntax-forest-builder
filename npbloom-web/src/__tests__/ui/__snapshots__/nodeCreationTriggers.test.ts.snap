// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`node creation triggers > generates node creation triggers for a tree with all slices assigned 1`] = `
[
  {
    "bottomRight": CoordsInTree {
      "treeX": 34,
      "treeY": -80,
    },
    "childIds": [
      "b",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 18,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 18,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 2,
      "treeY": -128,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 105.25,
      "treeY": -80,
    },
    "childIds": [
      "d",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 89.25,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 89.25,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 73.25,
      "treeY": -128,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 69.625,
      "treeY": -80,
    },
    "childIds": [
      "b",
      "d",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 18,
        "treeY": -60,
      },
      CoordsInTree {
        "treeX": 89.25,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 53.625,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 37.625,
      "treeY": -128,
    },
  },
]
`;

exports[`node creation triggers > generates node creation triggers for a tree with unassigned slices 1`] = `
[
  {
    "bottomRight": CoordsInTree {
      "treeX": 34,
      "treeY": -80,
    },
    "childIds": [
      "b",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 18,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 18,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 2,
      "treeY": -128,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 73,
      "treeY": -80,
    },
    "childIds": [
      "d",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 57,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 57,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 41,
      "treeY": -128,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 53.5,
      "treeY": -80,
    },
    "childIds": [
      "b",
      "d",
    ],
    "childPositions": [
      CoordsInTree {
        "treeX": 18,
        "treeY": -60,
      },
      CoordsInTree {
        "treeX": 57,
        "treeY": -60,
      },
    ],
    "origin": CoordsInTree {
      "treeX": 37.5,
      "treeY": -100,
    },
    "topLeft": CoordsInTree {
      "treeX": 21.5,
      "treeY": -128,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 108,
      "treeY": 0,
    },
    "origin": CoordsInTree {
      "treeX": 92,
      "treeY": -20,
    },
    "slice": StringSlice {
      "endExclusive": 15,
      "isZeroLength": false,
      "start": 11,
    },
    "topLeft": CoordsInTree {
      "treeX": 76,
      "treeY": -48,
    },
  },
  {
    "bottomRight": CoordsInTree {
      "treeX": 153.5,
      "treeY": 0,
    },
    "origin": CoordsInTree {
      "treeX": 137.5,
      "treeY": -20,
    },
    "slice": StringSlice {
      "endExclusive": 24,
      "isZeroLength": false,
      "start": 16,
    },
    "topLeft": CoordsInTree {
      "treeX": 121.5,
      "treeY": -48,
    },
  },
]
`;
